lib/vnet/vnet_test.go

```diff
@@ -0,0 +1,281 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package vnet
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"io"
+	"log/slog"
+	"net"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"testing"
+
+	"github.com/gravitational/trace"
+	"github.com/stretchr/testify/require"
+	"gvisor.dev/gvisor/pkg/tcpip"
+	"gvisor.dev/gvisor/pkg/tcpip/adapters/gonet"
+	"gvisor.dev/gvisor/pkg/tcpip/link/channel"
+	"gvisor.dev/gvisor/pkg/tcpip/network/ipv6"
+	"gvisor.dev/gvisor/pkg/tcpip/stack"
+
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+func TestMain(m *testing.M) {
+	utils.InitLogger(utils.LoggingForCLI, slog.LevelDebug)
+	os.Exit(m.Run())
+}
+
+type testPack struct {
+	vnetIPv6Prefix tcpip.Address
+	manager        *Manager
+
+	testStack        *stack.Stack
+	testLinkEndpoint *channel.Endpoint
+	localAddress     tcpip.Address
+}
+
+func newTestPack(t *testing.T, ctx context.Context) *testPack {
+	// Create two sides of an emulated TUN interface: writes to one can be read on the other, and vice versa.
+	tun1, tun2 := newSplitTUN()
+
+	// Create an isolated userspace networking stack that can be manipulated from test code. It will be
+	// connected to the VNet over the TUN interface. This emulates the host networking stack.
```

Comment:
nit: this comment makes it sound like a test-specific thing, but in fact this is exactly what we're using in production

Reply:
it is a test-specific thing, update the comment:

```
// Create an isolated userspace networking stack that can be manipulated from test code. It will be
// connected to the VNet over the TUN interface. This emulates the host networking stack.
// This is a completely separate gvisor network stack than the one that will be created in NewManager -
// the two will be connected over a fake TUN interface. This exists so that the test can setup IP routes
// without affecting the host running the Test.
```